### PR TITLE
Use external prices for conversion of fees in autopilot

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Related to discussion in the issue: https://github.com/cowprotocol/services/issues/2862#issuecomment-2270980569

Use external price vector to convert "fee in surplus token" into "fee in sell token" (instead of uniform clearing prices).

This is important so that we make sure that it's always true:

`total_fee = executed_surplus_fee + protocol fees`

If we use uniform clearing prices for `executed_surplus_fee` and external prices for `protocol fees` I think above equation won't stand.

## How to test
Existing unit test

### CODE REVIEW

File: crates/autopilot/src/domain/settlement/solution/mod.rs [hunk 1]
```diff
@@ -307,9 +310,15 @@ mod tests {
             eth::U256::from(52937525819789126u128)
         );
         // fee read from "executedSurplusFee" https://api.cow.fi/mainnet/api/v1/orders/0x10dab31217bb6cc2ace0fe601c15d342f7626a1ee5ef0495449800e73156998740a50cf069e992aa4536211b23f286ef88752187ffffffff
+        // "executedSurplusFee" and native fee are equal because the sell token is ETH
         assert_eq!(
             solution.native_fee(&auction.prices).0,
-            eth::U256::from(6890975030480504u128)
+            eth::U256::from(6752697350740628u128)
```
Comment (sunce86) [inline crates/autopilot/src/domain/settlement/solution/mod.rs:316]: Here we see that for this specific example, difference is ~2% if we use external prices vector for conversion. Seems acceptable.

**Comment by sunce86** [comment]
@fhenneke FYI this change affects:
`settlement_observation::fee`
`order_execution::executed_surplus_fee`

I don't expect this to have any significant effect on solver scripts.

**Review by fleupold — APPROVED** [review]
We already trust external price vectors for the surplus computation, so I think it make sense to use them for fee computation as well.

**Review by squadgazzz — APPROVED** [review]
LGTM

**Review by Mateo-mro — APPROVED** [review]

**Comment by fhenneke** [comment]
This will probably impact slippage and/or solver rewards.

If we want to reconstruct network fees (in sell token and in ETH) given protocol fees (in surplus token) and the total fee (in ETH), would we need external prices for all buy and sell tokens?

**Comment by sunce86** [comment]
> If we want to reconstruct network fees (in sell token and in ETH) given protocol fees (in surplus token) and the total fee (in ETH), would we need external prices for all buy and sell tokens?

Not sure what's the context of the question but given the protocol fees in surplus token, you need at least external price of buy token if you want to convert protocol fee to ETH for sell orders.
And also, if you want to convert protocol fee to sell token for sell orders you also need external price for both sell and buy token.
So I think the answer is yes.

**Comment by sunce86** [comment]
I think we don't give strong guarantees for COW AMM orders:

1. We [try to get all prices ](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/solvable_orders.rs#L222-L223) but if some are missing we are not filtering cow amm order owners for those that are missing.
2. Reference driver would filter out solutions that miss external price but external driver who knows
3. Autopilot does not check if all orders in the `/solve` response have prices (at the competition time).

This can be resolved with https://github.com/cowprotocol/services/issues/2720

**Comment by fhenneke** [comment]
The slippage query will have to be adapted (it uses surplus_fee for exactly computing network fees) and we might need to upload additional data to dune. At the moment, external prices are uploaded for protocol fee tokens only. With this PR we need it for sell and buy tokens.

The midterm plan is to compute slippage off-chain and then it does not make as much of a difference.

**Comment by fleupold** [comment]
> (this is a bug maybe?).

yes, if we cannot compute native prices the AMM should not tradable because we won't be able to compute surplus. That's a bug.

**Pull request merged**